Sources/GRPC/Interceptor/ClientInterceptorContext.swift

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright 2020, gRPC Authors All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import Logging
+import NIO
+
+public struct ClientInterceptorContext<Request, Response> {
+  /// The pipeline this context is associated with.
+  private let pipeline: ClientInterceptorPipeline<Request, Response>
+
+  /// The index of this context's interceptor within the pipeline.
+  private let index: Int
+
+  // The next context in the inbound direction, if one exists.
+  private var nextInbound: ClientInterceptorContext<Request, Response>? {
+    return self.pipeline.context(atIndex: self.index + 1)
+  }
+
+  // The next context in the outbound direction, if one exists.
+  private var nextOutbound: ClientInterceptorContext<Request, Response>? {
+    return self.pipeline.context(atIndex: self.index - 1)
+  }
+
+  /// The `EventLoop` this interceptor pipeline is being executed on.
+  public var eventLoop: EventLoop {
+    return self.pipeline.eventLoop
+  }
+
+  /// A logger.
+  public var logger: Logger {
+    return self.pipeline.logger
+  }
+
+  /// Construct a `ClientInterceptorContext` for the interceptor at the given index within in
+  /// interceptor pipeline.
+  internal init(pipeline: ClientInterceptorPipeline<Request, Response>, index: Int) {
+    self.pipeline = pipeline
+    self.index = index
+  }
+
+  /// Forwards the response part to the next inbound interceptor in the pipeline, if there is one.
+  ///
+  /// - Parameter part: The response part to forward.
+  /// - Note: This does *not* have to be called from the `EventLoop`.
```

Comment:
Is that worthwhile? NIO doesn't bother with this.

Reply:
Hmm I figured it'd be better to avoid failure since this could be called from off of the event loop, e.g. after deferring to e.g. UI for authentication or something. Perhaps it'd be better to document that it must be called from the event loop and keep the relevant assertions. WDYT?

Reply:
Yeah, I think that that's probably better? It's a pretty significant perf win to avoid the check.

Reply:
I gave the pipeline the same treatment: it'll be internal and the inbound side will already be on the event loop. The outbound side will be invoked from the call object, so we can push the check there instead.